proposals/0088-enable-core-bpf-programs.md

```diff
@@ -0,0 +1,143 @@
+process.
+
+With an explicit process in place for upgrading Core BPF programs, engineers can
+safely manage changes to essential programs through feature gates. The feature
+gate process requries incremental activations on each cluster: testnet, then
+devnet, then mainnet-beta. This is extremely valuable for ensuring a change has
+been safely integrated and will not have negative effects on mainnet-beta.
+
+This same explicit process can be used to migrate existing native programs to
+their new Core BPF versions.
+
+## Alternatives Considered
+
+An alternative approach to managing upgrades to essential core programs could be
+to make these core BPF programs _upgradeable_ and establish a multi-signature
+upgrade authority. This authority could be comprised of key-holders from each
+validator client: Solana Labs, Jito, Jump, and possibly Syndica (Sig).
+
+However, this approach adds additional layers of complexity around upgrading
+these programs behind a feature-gate.
+
+## New Terminology
+
+- `Core BPF Program`: A **non-upgradeable** BPF program relied on by any part of
+  the Solana stack, including (but not limited to) consensus, transaction
+  processing, voting, staking, and account creation.
+- `Target Program`: The core program intending to be changed by some
+  feature activation.
+- `Source Program`: The modified version of a core program intending to be moved
+  in place of the existing core program.
+
+## Detailed Design
+
+Core BPF programs shall be non-upgradeable BPF programs deployed with
+`BPFLoader2111111111111111111111111111111111`.
+
+Core BPF programs must only be modified through feature gates.
+
+**Upgrading a core BPF program** shall consist of deploying the modified program
+source code to a new arbitrary address as a **non-upgradeable** BPF program and
+creating a feature gate to, using the runtime, move the modified program in
+place of the existing program. 
+
+Note, since this deployed program will be deployed using the non-upgradeable BPF
+loader (`BPFLoader2111111111111111111111111111111111`), it will consist of only
+a program account, with no additional program _data_ account.
+
+This process must adhere to the following steps:
+
+1. Verifiably build the modified program.
+2. Generate a new keypair for the **source** program.
+3. Deploy the program to the **source** address.
+4. Generate a new keypair for the **feature gate**.
+5. Create a new feature gate for replacing the **target** program with the
+   **source** (modified) program.
+6. Follow the existing process for activating features.
+
+An additional optional field could be added to feature gate issues for the
+**target program** being upgraded.
+
+**Migrating a native program** to a core BPF program shall consist of following
+the exact steps outlined above with the modified BPF version of the program.
+This will effectively replace the program at the target address with its core
+BPF implementation.
```

Comment:
Up to you, but I think there might be more to this since the runtime will also need to remove any special-casing for that address during transaction processing.

Reply:
I think it's worth including the additional steps in the SIMD. What does it look like to remove special casing?

Reply:
Since this SIMD is now slimmed down and mentions that upon any migration, a SIMD outlining the migration and Core BPF specifications should be published, I think describing special casing removal isn't necessary here.

However, I've added it to the necessary details for such a SIMD as mentioned in this proposal.